### ISSUE TITLE
[RFC] Less user-hostile GregTech GUIs

### ISSUE THREAD

**Issue by YannickMG** [GTNewHorizons/GT-New-Horizons-Modpack#9367]
### Your GTNH Discord Username

Yannick#5937

### Your Pack Version

n/a

## Your Proposal

TLDR; I'm currently working on implementing this. Because this change could have a big impact on everyone's experience, please tell me (constructively) what you dislike about the current proposal and implementation, what parts of it you'd like to see improved/toggleable. Feedback on the (poor) visuals, on  the tooltip language and other implementation details are all welcome! Ideas for future uses of tooltip and tab capabilities probably belong in **a separate feature request**.

### Preface

Provide users with a clear, informative and usable interface, using tooltips, tabs, consistent design language so that they can focus on solving fun problems, not on fighting the UI.

My proposal for ***beginning*** to address these issues involves involves adding 2 configurable features to GregTech UIs: Enhanced tooltips and tabs.

## PART 1 : Tooltips

Any element of the UI which doesn't have a clear and unambiguous signification should have an accompanying tooltip that explains how to use it. This includes explaining the purpose of seemingly unused slots, exposing hidden interactions with items (clicking  fluid slots with cells for example) and hidden useful information (machine tank capacity) as well as explaining the few errors displayed on the UI. Ideally all Gregtech GUIs would eventually benefit from this, but this proposal focuses on Basic Machine GUIs for the time being.

### Configuring tooltips
I propose 4 different tooltip configurations:
- Disabled : No additional tooltips are added anywhere
- Toggleable : A UI element allows turning on extra tooltips when you want them to be, and leaves them off the rest of the time
- Toggleable with critical tooltips: Same as Toggleable except tooltips marked as "critical" such as explanations of machine errors are always enabled.
- Enabled : All new tooltips are always enabled.

I would love to hear which one you'd rather have be the default.

#### Toggleable tooltips
![tooltip-toggle](https://user-images.githubusercontent.com/5321549/148161009-9ede1ae3-d76b-4239-85f1-94aa56ac8277.gif)

#### Blocked steam vent
![image](https://user-images.githubusercontent.com/5321549/148161095-395c649f-309e-4289-ac49-844425d586c3.png)

#### Insufficient steam
![image](https://user-images.githubusercontent.com/5321549/148161111-bbe25929-a03a-4b9f-8a84-9d99d8bc8d4e.png)

#### Power Slot
![image](https://user-images.githubusercontent.com/5321549/148290425-e6ebf113-d9e3-43e7-b801-b551a862a95a.png)

#### Special Slot
![image](https://user-images.githubusercontent.com/5321549/148290469-3c58622d-96df-491c-9967-cc46b340da6c.png)
![image](https://user-images.githubusercontent.com/5321549/148290572-8fb92c2c-ec4e-4ae3-bc24-9c9f05fa243b.png)

#### Fluid Input Slot
![image](https://user-images.githubusercontent.com/5321549/148290669-57f3f6ba-da2a-4dc1-bf23-0bce4396409c.png)

#### Fluid Output Slot
![image](https://user-images.githubusercontent.com/5321549/148290908-d200e65a-7203-46da-83a3-c1ed14670f13.png)

## PART 2 : Tabs

There is simply not enough available space on the current machine GUIs to for all of the information and interactivity that should be there. A popular solution in many mods is to add tabs to the side of the UI which host additional functionality and information.

All that being said Gregtech already has a pretty unique way to extend the usability and capabilities of machines in the form of covers! Rather than try and replace covers, I felt it would be more appropriate to expand their uses and usability by giving them a place in the main machine UI as one of the first use of tabs.

Future possible use of tabs include showing additional information and exposing hidden machine configuration options.

### Configuring tabs

I think that tabs are too powerful a UI capability to be completely turned off. This being said one should be able to opt out of specific features implemented using tabs as long as there's another way to access said feature.

For that purpose I propose a configuration option to turn off the specific features:
- Cover tabs
- Tooltip toggle tab (Can be disabled by not using a Toggleable configuration option for tooltips

### Tabs in action
![cover-tabs-demo-cut](https://user-images.githubusercontent.com/5321549/148291228-3cf470b2-027d-44a7-bd11-97e7c3fd1130.gif)

## Your Goal

Currently the GUI for GregTech machines does a very poor job at actually informing you of what the machine's capabilities are, how the machine works and why it fails to work so much so that I'd label it as being user-hostile (as opposed to user-friendly). Error messages only exist in mutltiblocks, some features are hidden entirely behind outside interactions with tools (like the screwdrive to toggle accepting inputs through the output side).

This all contributes to making the learning curve of the basic interface of the mod much higher than it needs to be. Deciphering the (lack of) UI is not the type of fun challenge this pack is about.

## Your Vision

Revamping the existing GUIs to be less user-hostile and to make covers easier to use could make it easier and more likely to see even further improvements in this direction moving forward.

They could even be used to move certain GUI elements (such as auto-output toggles) off to the side, freeing up more space for additional fluid input/output slots on single-block machines!

## Final Checklist

- [X] I have searched this issue tracker and there is nothing similar already. Posting on a closed issue saying *I like this change please reconsider adding it* will prompt us to investigate and reopen it once we confirm your report.
- [X] I understand this change request may not attract enough attention and thus not be implemented.
- [X] I understand this change request may be rejected due to other community members think it's inappropriate.
- [X] I believe this feature would make the pack better.

**Comment by draknyte1** [comment]
Multi input tanks on machines requires a bit of overhaul and then more work when you want stuff in specific slots. I did it once for a dual fuel generator and it wasn’t worth the hassle. 

Tooltips and tabs should be disabled via mod options in game, but otherwise look neat.

**Comment by bombcar** [comment]
I think toggleable *via config* or via mod options is the way to go. I don't think wasting a tab on the ℹ️ button is a useful place for that toggle, as it's a set it forget it type of thing.

It would be *nice* if the tooltip setting was stored in player data instead of in the config, because either way the players would have to switch it back on each update (or meld) if whatever there preference was wasn't the default.

**Comment by YannickMG** [comment]
> Multi input tanks on machines requires a bit of overhaul and then more work when you want stuff in specific slots. I did it once for a dual fuel generator and it wasn’t worth the hassle.
> 
> Tooltips and tabs should be disabled via mod options in game, but otherwise look neat.

Thanks, any concerns/wishlist for addon support/overridability?

**Comment by GTNH-Colen** [comment]
Perhaps you could fold in #9360 and add gui support inside input busses for limiting and sorting modes? Especially if you are gonna add these features for single blocks at the same time. Two small buttons would work fine (maybe also the redstone output with the soldering alloy mode, but tbh idk what this feature does).

**Comment by YannickMG** [comment]
> Perhaps you could fold in #9360 and add gui support inside input busses for limiting and sorting modes? Especially if you are gonna add these features for single blocks at the same time. Two small buttons would work fine (maybe also the redstone output with the soldering alloy mode, but tbh idk what this feature does).

I'd love to do that as an additional feature which uses the same tab support implemented here, unless you can see some ways in which the proposed implementation should be modified to accommodate that future need.

**Comment by GTNH-Colen** [comment]
Should be fine, two buttons from the top left leaves plenty of space for any future features we add.

**Comment by Minepolz320** [comment]
Amazing job!

**Comment by YannickMG** [comment]
Since I linked to this from Discord, let me just add that in https://github.com/GTNewHorizons/GT5-Unofficial/pull/858 the tooltip toggle idea was changed for LSHIFT to toggle instead of a button, with configurable verbosity levels.

**Comment by YannickMG** [comment]
> Things that would be nice to have: gui buttons for toggling allow input on the output face (currently modified with rclick output or face with screwdriver) gui button for toggling the state of multi-stack input (currently modified with shift rclick face with screwdriver) gui button for toggling the state of input filter (currently modified with shift rclick face with screwdriver) gui button for toggling disabled state (currently modified with soft mallet)

Hey Fourk, while this is all possible and something I'd like to do, do you think you can make a new separate Feature Request issue with these suggestion?

**Comment by MadRatSRP** [comment]
I also want to add a few suggestion which applies both for GUI and tooltips:
1)Vertical or horizontal bar for all of the machines which shows the amount of steam/energy which this machine currently contains
2)In some quests of steam age some tooltips of the machines shows the amount of steam this machine uses and i think it should be in general like that. In tooltip of the block of the machines/generators there should be information about how much energy/steam does this machine produces/consumes

**Comment by leagris** [comment]
> Things that would be nice to have: gui buttons for toggling allow input on the output face (currently modified with rclick output or face with screwdriver) gui button for toggling the state of multi-stack input (currently modified with shift rclick face with screwdriver) gui button for toggling the state of input filter (currently modified with shift rclick face with screwdriver) gui button for toggling disabled state (currently modified with soft mallet)

I think it is a user hostile, conter-intuitive to involve tools in GUI interactions.
GUI are purely virtual stuffs to interface real-world with in-game interactions.
GUI are here to provide player↔game service. This they shall not involve in-game tools.

If you really insist on adding tools to GUI controls, these shall not incur damage to tools at all.

If you were to give the user a better clue of the interaction type on a given GUI widget like a screwdriver control or a soft-mallet push-button operation control. Then you should change the mouse pointer icon respectively to a screwdriver or a soft-mallet icon.
These are client-side and user-side control elements that belongs to the GUI and shall absolutely not be in-game elements.

**Comment by YannickMG** [comment]
> > Things that would be nice to have: gui buttons for toggling allow input on the output face (currently modified with rclick output or face with screwdriver) gui button for toggling the state of multi-stack input (currently modified with shift rclick face with screwdriver) gui button for toggling the state of input filter (currently modified with shift rclick face with screwdriver) gui button for toggling disabled state (currently modified with soft mallet)
> 
> I think it is a user hostile, conter-intuitive to involve tools in GUI interactions. GUI are purely virtual stuffs to interface real-world with in-game interactions. GUI are here to provide player↔game service. This they shall not involve in-game tools.
> 
> If you really insist on adding tools to GUI controls, these shall not incur damage to tools at all.
> 
> If you were to give the user a better clue of the interaction type on a given GUI widget like a screwdriver control or a soft-mallet push-button operation control. Then you should change the mouse pointer icon respectively to a screwdriver or a soft-mallet icon. These are client-side and user-side control elements that belongs to the GUI and shall absolutely not be in-game elements.

Then again, the maintenance hatch is a thing.
